DOC: developer docs for plugin components (#1263)
spacetelescope/jdaviz: docs/dev/ui_plugin_components.rst

Design Philosophy
-----------------

Each component consists of three parts:
1. Python component class in ``template_mixin.py`` which inherits from ``BasePluginComponent`` and passes the string names of the traitlets it needs to `__init__` as kwargs.  This class isolates the logic from the plugin itself, while still providing convenient access to the traitlets which are defined in the plugin itself (and in doing so, allows using those same traitlets within the plugin in the same way as any other traitlet in the plugin).  Within this class, it is necessary to use ``self.add_observe`` instead of the ``@observe`` decorator for all traitlets, so that the callback can reference the traitlet in the plugin properly.
2. Python mixin class in ``template_mixin.py`` which inherits from ``VuetifyTemplate`` and ``HubListener``.  This class defines default traitlets as well as the attribute for the component object itself for plugins to make use of the accompanying component.  In some cases, the component class will be used manually with custom traitlets (especially if/when using multiple instances of the same component within a single plugin).

> 3. `.vue` template file in ``jdaviz/components``, which are registered in ``app.py``.  These templates are not linked directly to the python class, but rather should pass all necessary traitlets and options when called from within the template file for the plugin itself.  Note that this means that the instance of the component cannot be rendered individually, but also allows for components to interact with each other easily through traitlet events within the plugin.  If nesting these inside each other, it might be necessary to manually re-emit events higher up the tree with something like ``@update:value="$emit('upadte_value', $event)"``.

Reviewer: Found a typo
```suggestion
3. `.vue` template file in ``jdaviz/components``, which are registered in ``app.py``.  These templates are not linked directly to the python class, but rather should pass all necessary traitlets and options when called from within the template file for the plugin itself.  Note that this means that the instance of the component cannot be rendered individually, but also allows for components to interact with each other easily through traitlet events within the plugin.  If nesting these inside each other, it might be necessary to manually re-emit events higher up the tree with something like ``@update:value="$emit('update_value', $event)"``.
```

*****************
Plugin Components
*****************

Plugin components exist to provide re-usable UI elements across multiple plugins, both for 
consistency in behavior between plugins and also for simplification of code.

The general concept is to move as much shared code into these components, and out of the plugins, as
possible.

Design Philosophy
-----------------

Each component consists of three parts: 

1. Python component class in ``template_mixin.py`` which inherits from ``BasePluginComponent`` and
passes the string names of the traitlets it needs to ``__init__`` as kwargs.  This class isolates the
logic from the plugin itself, while still providing convenient access to the traitlets which are
defined in the plugin itself (and in doing so, allows using those same traitlets within the plugin
in the same way as any other traitlet in the plugin). Within this class, it is necessary to use
``self.add_observe`` instead of the ``@observe`` decorator for all traitlets, so that the callback

> can reference the traitlet in the plugin properly. 

Reviewer: If you have a better example, feel free to replace suggestion.
```suggestion
can reference the traitlet in the plugin properly.
For example implementation, see `~jdaviz.core.template_mixin.BaseSelectPluginComponent`.
```

2. Python mixin class in ``template_mixin.py`` which inherits from ``VuetifyTemplate`` and
``HubListener``. This class defines default traitlets as well as the attribute for the component
object itself for plugins to make use of the accompanying component.  In some cases, the component
class will be used manually with custom traitlets (especially if/when using multiple instances of
the same component within a single plugin). 

> 3. ``.vue`` template file in ``jdaviz/components``, which are registered in ``app.py``.  These

Reviewer: ```suggestion
3. ``.vue`` template file in ``jdaviz/components``, which are registered in ``app.py``
using ``ipyvue.register_component_from_file()``.  These
```

> templates are not linked directly to the python class, but rather should pass all necessary

Reviewer: ```suggestion
templates are not linked directly to the Python class, but rather should pass all necessary
```

traitlets and options when called from within the template file for the plugin itself. Note that
this means that the instance of the component cannot be rendered individually, but also allows for
components to interact with each other easily through traitlet events within the plugin. If nesting
these inside each other, it might be necessary to manually re-emit events higher up the tree with

> something like ``@update:value="$emit('update_value', $event)"``.

Reviewer: ```suggestion
something like ``@update:value="$emit('update_value', $event)"`` in the relevant ``.vue`` file.
```

> ``BasePluginComponent`` provides the following functionality to all components:

Reviewer: ```suggestion
`~jdaviz.core.template_mixin.BasePluginComponent` provides the following functionality to all components:
```

> * ``app``, ``hub``, and ``plugin`` properties to access the respective instances.

Reviewer: I don't see any public `.plugin` defined in the code. Do you mean `._plugin` that is "hidden"?
```suggestion
* `~jdaviz.core.template_mixin.BasePluginComponent.app`, `~jdaviz.core.template_mixin.BasePluginComponent.hub`,
  and ``plugin`` properties to access the respective instances.
```

Also you might want to add a concise docstring for the public properties mentioned here.

> * ``viewer_dicts`` property to access a list of dictionaries, with keys: ``viewer``, ``id``, 

Reviewer: ```suggestion
* `~jdaviz.core.template_mixin.BasePluginComponent.viewer_dicts` property
  to access a list of dictionaries, with keys: ``viewer``, ``id``, 
```

  ``reference``, and ``label`` (``reference`` if available, otherwise ``id``).

> * ``add_observe`` function to connect a callback to the proper traitlet in the parent plugin.  This

Reviewer: ```suggestion
* :meth:`~jdaviz.core.template_mixin.BasePluginComponent.add_observe` method to
  connect a callback to the proper traitlet in the parent plugin. This
```

  can optionally take ``first=True`` to ensure the callback will be processed before any
  ``@observe`` watchers in the plugin.
* overrides ``getattr`` and ``setattr`` to redirect any calls to the internal traitlet attributes
  to those in the plugin.

> Considerations

Reviewer: It seems to me this section is mixing up both pros and cons? Given that I am very unfamiliar with the innerworkings here, I am not sure if the points are "things that will hurt me if I am not careful" or "things that I do not have to worry about if I use the stuff mentioned".

If indeed this is a mix of both, it is better to separate them.

--------------

> * Each class can only subscribe to each ``Message`` object once (via ``self.hub.subscribe``),

Reviewer: ```suggestion
* Each class can only subscribe to each `~glue.core.message.Message` object once (via ``self.hub.subscribe``),
```

  without introducing extra wrappers.  By isolating the component-logic from the plugin, the
  component and the plugin itself can subscribe to the same message without issues.

> * Having the whole component in a Mixin would also prevent the ability to have multiple instances

Reviewer: Is this what you really mean? Please correct if I misunderstood.
```suggestion
* Having the whole component inheriting from a mixin would also prevent the ability to have multiple instances
```

Author: this isn't quite what I mean.  Would "Having all the logic in a Mixin, instead of a Mixin wrapping around a separate "component" class", be more clear?

>   of the same component within a single plugin.  This is needed in several places: subsets in line

Reviewer: Maybe explain a little *why* this was needed, say, in line analysis; e.g., is it to prevent component from going out of sync under certain workflow?
